### PR TITLE
Fixes angle axis to quaternion order bug

### PR DESCRIPTION
**Pull request by amonszpart**
### Description
- `angle_axis_to_quaternion` and `quaternion_to_angle_axis` assumed quaternion coefficients to be in `(w, x, y , z)` order (real part first), whilst the rest of the functions working with quaternions, and the function documentation assumes quaternions to be real part last, i.e. `(x, y, z, w)`. 
- This PR fixes `angle_axis_to_quaternion` and `quaternion_to_angle_axis` by rolling the indices around.
- A few new integration tests were also created, that check for the compatibility among `quaternion_to_...`, `rotation_matrix_to_...` and `angle_axis_to_...` functions. Tests like this would've caught the errors above.

### Status
Ready

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New tests added to cover the changes

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [x] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/geometry/conversions.py [hunk 1]
```diff
@@ -679,7 +679,7 @@ def angle_axis_to_quaternion(angle_axis: torch.Tensor) -> torch.Tensor:
     quaternion[..., 0:1] += a0 * k
     quaternion[..., 1:2] += a1 * k
     quaternion[..., 2:3] += a2 * k
-    return torch.cat([w, quaternion], dim=-1)
+    return torch.cat([quaternion, w], dim=-1)
```
Comment (edgarriba) [inline kornia/geometry/conversions.py:682]: if we go for [w, x, y, z] I guess we shouldn't modify this.

**Review by edgarriba — COMMENTED** [review]
depending the decision in https://github.com/kornia/kornia/issues/903 I think we should revisit other functions so verify consistency over the other functionality that involve quaternion.

File: kornia/geometry/conversions.py (hunk 1 above)
Comment (amonszpart) [reply kornia/geometry/conversions.py:682]: Quite the feature creep...how do you propose to warn against backward compatibility?
Something like 
```python
def quaternion_to_rotation_matrix(quaternion: torch.Tensor, order: str = 'xyzw') -> torch.Tensor:
    if order not in ('wxyz', 'xyzw'):
        raise RuntimeError(f"Need order to be one of {'xyzw', 'wxyz'}")
   if order == 'xyzw':
        logger.warning(f'Order xyzw deprecated, please use wxyz')
```
?

Comment (edgarriba) [reply kornia/geometry/conversions.py:682]: that's an option, yes

**Comment by edgarriba** [comment]
@amonszpart thanks for the PR - crazy number of tests you added :) I'll check in detail over the week. Seems that integration tests still failing.

File: kornia/geometry/conversions.py (hunk 1 above)
Comment (amonszpart) [reply kornia/geometry/conversions.py]: Ok, I think it's ready for review again. Not pretty, but transitions never are.

**Comment by edgarriba** [comment]
@amonszpart I'll take this today and fix the issues with jit

**Comment by amonszpart** [comment]
> @amonszpart I'll take this today and fix the issues with jit

Thanks. Most of my tests were passing, except PyTorch 1.6 a few commits ago, so not sure what went wrong with my last commit. If there is something systematic you need me to change, let me know. I just can't parse what GitHub is outputting right now. 

Also, how do you add linebreaks to example usage at the end of a docstring? I tried just "enter",  appending `\` and not breaking the line, but neither worked...

**Comment by shijianjian** [comment]
> Also, how do you add linebreaks to example usage at the end of a docstring? I tried just "enter",  appending `\` and not breaking the line, but neither worked...

You might use enter normally but adding "... " in the front instead of ">>> ".

**Comment by edgarriba** [comment]
@amonszpart the other issues I believe that are because torchscript doesn't like uninitialized variables

File: kornia/geometry/conversions.py [hunk 2]
```diff
@@ -340,16 +354,22 @@ def rotation_matrix_to_quaternion(
...
-            type(rotation_matrix)))
+        raise TypeError(f"Input type is not a torch.Tensor. Got {type(rotation_matrix)}")
 
     if not rotation_matrix.shape[-2:] == (3, 3):
-        raise ValueError(
-            "Input size must be a (*, 3, 3) tensor. Got {}".format(
-                rotation_matrix.shape))
+        raise ValueError(f"Input size must be a (*, 3, 3) tensor. Got {rotation_matrix.shape}")
+
+    if order not in QuaternionCoeffOrder.ALL:
```
Comment (edgarriba) [inline kornia/geometry/conversions.py:365]: @amonszpart @shijianjian apparently jit doesn't like this comparison :)

File: kornia/geometry/conversions.py [hunk 3]
```diff
@@ -340,16 +354,22 @@ def rotation_matrix_to_quaternion(
...
+        raise ValueError(f"order must be one of {QuaternionCoeffOrder.ALL}")
+
+    if order == QuaternionCoeffOrder.XYZW:
```
Comment (edgarriba) [inline kornia/geometry/conversions.py:368]: same for this

**Comment by amonszpart** [comment]
> @amonszpart the other issues I believe that are because torchscript doesn't like uninitialized variables

Ok, so that means I have to drop the type annotation, or there'll be a fail with mypy because of double type annotation.

File: kornia/geometry/conversions.py (hunk 2 above)
Comment (amonszpart) [reply kornia/geometry/conversions.py:365]: So I can change from .ALL to != WXYZ, and ~= XYZW, but it still complains because of the str value of the non Enum. 
I have very little experience with JIT, so not really the person to fix this...
I tried `global QuaternionCoeffOrder`, but JIT doesn't like globals. 

```shell
E           RuntimeError: 
E           python value of type 'str' cannot be used as a value. Perhaps it is a closed over global variable? If so, please consider passing it in as an argument or use a local varible instead.:
E             File "/home/aron/workspace/kornia/kornia/geometry/conversions.py", line 639
E                   raise ValueError(f"Input must be a tensor of shape (*, 3). Got {quaternion.shape}")
E           
E               if order != QuaternionCoeffOrder.XYZW and order != QuaternionCoeffOrder.WXYZ:
E                           ~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
E                   raise ValueError(f"order must be one of {QuaternionCoeffOrder.ALL}")

../../.dev_env/envs/venv/lib/python3.7/site-packages/torch/jit/_script.py:990: RuntimeError
```

**Comment by edgarriba** [comment]
@amonszpart thanks for updating. I need to check our CI with cuda since has been giving memory errors recently. as for torchscript since to fail only with PT 1.6 , maybe could skip tests on that version ?

**Comment by amonszpart** [comment]
@edgarriba done! :)
